Added support for authenticating with SASL-SCRAM:

squ1d123 opened the pull request:
- added secrets file for kafka and zookeeper auth
 - updated values.yaml with config for sasl-scram
 - add config-map for jaas file to use for java authentication
 - added KAFKA_OPTS environment variable for specifying location of jaas file
 - added init container to substitute secrets into jaas file before application startup

<!--
Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

**What this PR does / why we need it**:
This enables the schema-registry to authenticate with kafka clusters using SASL-SCRAM.
It does so by creating kubernetes secrets and injecting them into tempfs in the form of a jaas file for the schema registry to use, therefore the secrets are only available in memory.

Default behaviour has not been changed.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

Commit:
Added support for authenticating with SASL-SCRAM:
 - added secrets file for kafka and zookeeper auth
 - updated values.yaml with config for sasl-scram
 - add config-map for jaas file to use for java authentication
 - added KAFKA_OPTS environment variable for specifying location of jaas
 file
 - added init container to substitute secrets into jaas file before
 application startup
 - bumped minor version as this is adding a new feature

squ1d123 commented:
Hey can this please be reviewed?
I know its a decent size change, however sasl-scram requires quite a few additions to get working.
/assign @benjigoldberg

benjigoldberg commented on incubator/schema-registry/templates/deployment.yaml:
```diff
@@ -15,6 +15,39 @@ spec:
+        - name: jaasconfig
+          mountPath: {{ .Values.sasl.configPath | quote }}
+    {{- end }}
+
```
rm extra newline

benjigoldberg commented on incubator/schema-registry/templates/sasl-scram-config-map.yaml:
```diff
@@ -0,0 +1,30 @@
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+
+  {{/* - range $i, $e := until (int .Values.replicas) */}}
```
Why is a range needed?

Also - spacing seems off here for your range

squ1d123 replied:
Yep, not needed here, was working on the `kafka` chart, and potentially creating a jaas file per broker with unique Client user/pass for authentication with zookeeper. Will remove the range.

benjigoldberg commented on incubator/schema-registry/values.yaml:
```diff
@@ -34,6 +34,23 @@ resources: {}
 ## The port on which the SchemaRegistry will be available and serving requests
 servicePort: 8081
 
+## Additional Java arguments to pass to Kafka.
```
Do you mean to define these as `schemaRegistryOpts` instead? If you want to configure the kafka subchart, these settings should go under `kafka`

squ1d123 replied:
Yes you are correct, I am also working on the Kafka helm chart, which is why I had this mixed up, will change to `schemaRegistryOpts`

benjigoldberg commented on incubator/schema-registry/values.yaml:
```diff
@@ -34,6 +34,23 @@ resources: {}
+      image: "confluentinc/cp-schema-registry"
+      imageTag: "4.0.0"
+      imagePullPolicy: "IfNotPresent"
+    clientUser: "kafka-client"
```
I think there should be an option with the secrets config where the user can specify the name of an already existing secret in Kubernetes for consumption. We should definitely allow the user to optionally generate the secret with the chart if desired, but in some environments/installations it will be preferable to have the ability to specify a secret separately.

squ1d123 replied:
Yep that's a fair point and I agree with you. I will add an option for the user to specify if they want to use their own secret or not, and the path to their own secret if opting to use their own.

benjigoldberg reviewed (changes requested):
This is a really great enhancement to the chart @squ1d123 -- Handful of comments but a great start!

Commit:
modifications based on pull requesrt feedback:
- implemented allowing the user to specify their own k8s secret
- changed kafkaOpts to schemaRegistryOpts
- cleaned up whitespace
- removed redundant range loop

Commit:
bumping minor version as this is adding a new feature

squ1d123 commented:
@benjigoldberg thanks for the feedback, I have made the fixes recommended, could you please review again. Thanks

timstoop commented:
/ok-to-test
Looks good to me, but @benjigoldberg will know best :)

benjigoldberg commented:
/lgtm
